Kernel_23/include/CGAL/Kernel/function_objects.h

```diff
@@ -2881,6 +2881,28 @@ namespace CommonKernelFunctors {
       return closest_point_on_segment;
     }
 
+    typename K::Point_3
+    operator()(const typename K::Point_3& query,
+               const typename K::Ray_3& ray,
+               const K& k)
+    {
+      typedef typename K::Point_3 Point_3;
+      
+      typename K::Construct_projected_point_3 projection =
+          k.construct_projected_point_3_object();
+      
+      
+      
+      // Project query on segment supporting line
+      const Point_3 closest_point = projection(ray.supporting_line(), query);
+      
+      if ( ray.to_vector() * (closest_point-ray.source()) < 0)
+        closest_point = ray.source();
+      
+      //returns the constructed point
+      return closest_point;
+    }
```

Comment:
That code is not optimal. No need to construct the projection point before testing the scalar product:
```C++
if ( ray.to_vector() * (query-ray.source()) < 0)
  return ray.source();
else
  return projection(ray.supporting_line(), query);
```

Reply:
What do you mean? the scalar product is using the projection.

Reply:
By definition of the projection, the scalar product with the projected point `p'` or with the initial query point `p` is the same. Let's say the source of the ray is `s` and its vector is `v`:

    p' = p + pp'
    sp'. v = sp . v + pp'. v

as `pp'` is the projection vector, it is orthogonal to `v`, so `pp' . v = 0`, and thus

    sp'. v = sp . v

Reply:
No, it constructs the projection, and then checks if it is inside, with two scalar products.
https://github.com/CGAL/cgal/blob/bcfb7afead5115415168fae5c5dac8dd40a06b94/Kernel_23/include/CGAL/Kernel/function_objects.h#L2853-L2882

Reply:
@maxGimeno it would be nice to fix it as the same time.